nixos/modules/services/security/kanidm.nix

```diff
@@ -62,41 +98,378 @@ let
+        exit 1
+      fi
+      [[ -n "$x" ]]
+    }
+
+    ${concatMapStrings (x: x._script) (attrValues cfg.provision.groups)}
+    ${concatMapStrings (x: x._script) (attrValues cfg.provision.persons)}
+    ${concatMapStrings (x: x._script) (attrValues cfg.provision.systems.oauth2)}
+
+    if [[ "''${needs_rewrite-0}" == 1 ]]; then
+      echo "Queueing service restart to rewrite secrets"
+      touch "$STATE_DIRECTORY/.needs_restart"
+    fi
+  '';
 in
 {
   options.services.kanidm = {
-    enableClient = lib.mkEnableOption (lib.mdDoc "the Kanidm client");
-    enableServer = lib.mkEnableOption (lib.mdDoc "the Kanidm server");
-    enablePam = lib.mkEnableOption (lib.mdDoc "the Kanidm PAM and NSS integration");
+    enableClient = mkEnableOption (mdDoc "the Kanidm client");
+    enableServer = mkEnableOption (mdDoc "the Kanidm server");
+    enablePam = mkEnableOption (mdDoc "the Kanidm PAM and NSS integration");
+
+    package = mkPackageOptionMD pkgs "kanidm" {};
+
+    provision = {
+      enable = mkEnableOption "provisioning of systems (oauth2), groups and users";
+
+      adminPasswordFile = mkOption {
+        description = mdDoc "Path to a file containing the admin password for kanidm. Do NOT use a file from the nix store here!";
+        example = "/run/secrets/kanidm-admin-password";
+        type = types.path;
+      };
+
+      idmAdminPasswordFile = mkOption {
+        description = mdDoc "Path to a file containing the idm admin password for kanidm. Do NOT use a file from the nix store here!";
+        example = "/run/secrets/kanidm-idm-admin-password";
+        type = types.path;
+      };
+
+      persons = mkOption {
+        description = mdDoc "Provisioning of kanidm persons";
+        default = {};
+        type = types.attrsOf (types.submodule (personSubmod: let
+          inherit (personSubmod.config._module.args) name;
+          updateArgs =
+            ["--displayname" personSubmod.config.displayName]
+            ++ optionals (personSubmod.config.legalName != null)
+            ["--legalname" personSubmod.config.legalName]
+            # mail addresses
+            ++ concatMap (addr: ["--mail" addr]) personSubmod.config.mailAddresses;
+        in {
+          options = {
+            _script = mkScript (
+              if personSubmod.config.present
+              then
+                ''
+                  if ! person_exists ${escapeShellArg name}; then
+                    kanidm-idm person create ${escapeShellArg name} \
+                      ${escapeShellArg personSubmod.config.displayName}
+                  fi
+                  kanidm-idm person update ${escapeShellArg name} ${escapeShellArgs updateArgs}
+                ''
+                + flip concatMapStrings personSubmod.config.groups (group: ''
+                  kanidm-idm group add-members ${escapeShellArg group} ${escapeShellArg name}
+                '')
+              else ''
+                if person_exists ${escapeShellArg name}; then
```

Comment:
If I'm reading this right, removing a person requires marking them as `present = false`. Can you instead make it so persons that are not defined at all get removed? (I consider that the fully declarative configuration style -- which is what sets NixOS apart from e.g. Ansible.)

Reply:
You are right you will need to specify `present = false` right now. In theory I could make it fully declarative but in practice this isn't feasible right now without writing an external tool that directly interacts with the kanidm API (which I might consider doing once the API is stable). I think it's too much work to bolt this on to the fragile and slow CLI, and this way it is at least flexible.

While I personally would like a fully declarative style here, I'm not sure that it will work properly for everyone in practice. As with all database systems you usually want to retain some kind of state and mutability at runtime. I don't know where this boundary will be and I don't know enough use cases to gauge this myself. It's hard to tell whether this would have a negative impact in some cases that I'm not aware of, especially since provisioning for other databases in nixpkgs works similarly.

Reply:
> I think it's too much work to bolt this on to the fragile and slow CLI, and this way it is at least flexible.

Ok. But why is fully declarative that much harder? List all users in database, remove the ones not listed in the config?

>While I personally would like a fully declarative style here, I'm not sure that it will work properly for everyone in practice.

We could have an option like `mutableUsers`? Fully declarative by default, opt-out to Ansible style declarativeness if desired, or no provisioning at all (upstream preferred way)?

Btw, I realize that https://github.com/glauth/glauth might be more aligned with "fully declarative" configuration. But kanidm seems so nice :-)

Reply:
> Ok. But why is fully declarative that much harder?

I didn't say hard. But tedious and prone to breakage. Something I'd rather do properly.

> We could have an option like `mutableUsers`? Fully declarative by default, opt-out to Ansible style declarativeness if desired, or no provisioning at all (upstream preferred way)?

Yeah something like that. I'll think about this more once the upstream API is stable enough.

> Btw, I realize that https://github.com/glauth/glauth might be more aligned with "fully declarative" configuration. But kanidm seems so nice :-)

I don't know it, but it seems very limited compared to what kanidm can do.

Reply:
Pushed an update right now, the new utility can track entities and thus `present = false;` is no longer required. Both explicit and implicit mode are supported and can be selected by specifying `autoRemove = true/false`.